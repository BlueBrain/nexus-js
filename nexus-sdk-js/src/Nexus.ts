import Store from './utils/Store';
import Organization, { OrganizationResponse } from './Organization';
import { httpGet, httpPost } from './utils/http';

type NexusConfig = {
  environment: string;
  token?: string;
};

export const store: Store = new Store({
  auth: {
    accessToken: null,
  },
  api: {
    baseUrl: null,
  },
});

export default class Nexus {
  constructor(config: NexusConfig) {
    if (!config.environment) {
      throw new Error(
        'No environment provided. Please specify your Nexus instance endpoint.'
      );
    }
    store.update('api', state => ({
      ...state,
      baseUrl: config.environment,
    }));
    if (config.token) {
      store.update('auth', state => ({
        ...state,
        accessToken: config.token,
      }));
    }
  }

  setToken(token: string): void {
    store.update('auth', state => ({
      ...state,
      accessToken: token,
    }));
  }

  removeToken(): void {
    store.update('auth', state => ({
      ...state,
      accessToken: undefined,
    }));
  }

  async listOrganizations(): Promise<Organization[]> {
    try {
      const orgs: OrganizationResponse[] = await httpGet('/orgs');
      return orgs.map((org: OrganizationResponse) => new Organization(org));
    } catch (e) {
      console.log(e);
      return e;
    }
  }

  async getOrganization(id: string): Promise<Organization> {
    try {
      const org: OrganizationResponse = await httpGet(`/orgs/${id}`);
      return new Organization(org);
    } catch (e) {
      return e;
    }
  }

  async createOrganization(): Promise<Organization> {
    try {
      const org: OrganizationResponse = await httpPost('/orgs', {});
      return new Organization(org);
    } catch (e) {
      return e;
    }
  }
}
