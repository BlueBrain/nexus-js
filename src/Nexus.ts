import store from './store';
import Organization from './Organization';
import { CreateOrgPayload, ListOrgOptions } from './Organization/types';
import makeOrgUtils, { OrgUtils } from './Organization/utils';
import { PaginatedList } from './utils/types';
import Store from './utils/Store';

type NexusConfig = {
  environment?: string;
  token?: string;
};

// Without this, jest crashes really bad...
if (process.env.NODE === 'TEST') {
  Organization;
}

export default class Nexus {
  readonly store: Store;
  readonly Organization: OrgUtils;

  // update global store
  static setEnvironment(environment: string): void {
    store.update('api', state => ({
      ...state,
      baseUrl: environment,
    }));
  }

  // update global store
  static setToken(token: string): void {
    if (!token || token === undefined || token.length === 0) {
      throw new Error('Token is invalid.');
    }
    store.update('auth', state => ({
      ...state,
      accessToken: token,
    }));
  }

  // update global store
  static removeToken(): void {
    store.update('auth', state => ({
      ...state,
      accessToken: undefined,
    }));
  }

  constructor(config: NexusConfig) {
    if (!config.environment) {
      throw new Error('No environment provided');
    }
    this.store = new Store({
      auth: { accessToken: config.token },
      api: { baseUrl: config.environment },
    });

    this.Organization = makeOrgUtils(this.store);
  }

  setToken(token: string) {
    if (!token || token === undefined || token.length === 0) {
      throw new Error('Token is invalid.');
    }
    this.store.update('auth', state => ({
      ...state,
      accessToken: token,
    }));
  }

  removeToken(): void {
    this.store.update('auth', state => ({
      ...state,
      accessToken: undefined,
    }));
  }

  async getOrganization(label: string): Promise<Organization> {
    return this.Organization.get(label);
  }

  async listOrganizations(
    listOrgOptions?: ListOrgOptions,
  ): Promise<PaginatedList<Organization>> {
    try {
      return this.Organization.list(listOrgOptions);
    } catch (error) {
      throw error;
    }
  }

  async createOrganization(
    label: string,
    orgPayload?: CreateOrgPayload,
  ): Promise<Organization> {
    try {
      return this.Organization.create(label, orgPayload);
    } catch (error) {
      throw error;
    }
  }

  async updateOrganization(
    label: string,
    rev: number,
    orgPayload: CreateOrgPayload,
  ): Promise<Organization> {
    try {
      return this.Organization.update(label, rev, orgPayload);
    } catch (error) {
      throw error;
    }
  }

  async deprecateOrganization(
    label: string,
    rev: number,
  ): Promise<Organization> {
    try {
      return this.Organization.deprecate(label, rev);
    } catch (error) {
      throw error;
    }
  }
}
