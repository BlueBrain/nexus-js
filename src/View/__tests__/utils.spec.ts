import { GlobalWithFetchMock } from 'jest-fetch-mock';
import Nexus from '../../Nexus';
import {
  listViews,
  getView,
  getElasticSearchView,
  getSparqlView,
} from '../utils';

const { fetchMock } = <GlobalWithFetchMock>global;
const baseUrl = 'http://api.url';
Nexus.setEnvironment(baseUrl);

const mockElasticViewResponse = {
  '@context': [
    'https://bluebrain.github.io/nexus/contexts/view.json',
    'https://bluebrain.github.io/nexus/contexts/resource.json',
  ],
  '@id': 'nxv:myview2',
  '@type': ['ElasticSearchView', 'View', 'Alpha'],
  includeMetadata: false,
  mapping: {
    dynamic: false,
    properties: {
      '@id': {
        type: 'keyword',
      },
      '@type': {
        type: 'keyword',
      },
      bool: {
        type: 'boolean',
      },
      name: {
        type: 'keyword',
      },
      number: {
        type: 'long',
      },
    },
  },
  resourceSchemas: ['https://bluebrain.github.io/nexus/schemas/myschema'],
  sourceAsText: false,
  _uuid: '4f90ceff-45b7-442d-8536-d41705321d50',
  _self: 'https://nexus.example.com/v1/schemas/myorg/myproj/nxv:myview',
  _constrainedBy: 'https://bluebrain.github.io/nexus/schemas/view',
  _project: 'https://nexus.example.com/v1/projects/myorg/myproj',
  _createdAt: '2018-09-18T16:39:22.748Z',
  _createdBy: 'https://nexus.example.com/v1/realms/myrealm/users/john',
  _updatedAt: '2018-09-18T17:10:22.748Z',
  _updatedBy: 'https://nexus.example.com/v1/realms/myrealm/users/john',
  _rev: 4,
  _deprecated: true,
};
const mockSparqlResponse = {
  '@context': [
    'https://bluebrain.github.io/nexus/contexts/view.json',
    'https://bluebrain.github.io/nexus/contexts/resource.json',
  ],
  '@id': 'nxv:myview',
  '@type': ['SparqlView', 'View', 'Alpha'],
  includeMetadata: false,
  resourceSchemas: ['https://bluebrain.github.io/nexus/schemas/myschema'],
  sourceAsText: false,
  _uuid: '4f90ceff-45b7-442d-8536-d41705321d50',
  _self: 'https://nexus.example.com/v1/schemas/myorg/myproj/nxv:myview',
  _constrainedBy: 'https://bluebrain.github.io/nexus/schemas/view',
  _project: 'https://nexus.example.com/v1/projects/myorg/myproj',
  _createdAt: '2018-09-18T16:39:22.748Z',
  _createdBy: 'https://nexus.example.com/v1/realms/myrealm/users/john',
  _updatedAt: '2018-09-18T17:10:22.748Z',
  _updatedBy: 'https://nexus.example.com/v1/realms/myrealm/users/john',
  _rev: 4,
  _deprecated: true,
};
const mockList = {
  _results: [mockElasticViewResponse, mockSparqlResponse],
  _total: 2,
};
const mockEmptyList = {
  _results: [],
  _total: 0,
};

describe('Views utils functions', () => {
  describe('getView()', () => {
    beforeEach(() => {
      fetchMock.mockResponse('{}', { status: 200 });
    });
    afterEach(() => {
      fetchMock.resetMocks();
    });

    it('should call the appropriate URL', async () => {
      await getView('myorg', 'myproject', 'viewid');
      expect(fetchMock.mock.calls[0][0]).toEqual(
        `${baseUrl}/views/myorg/myproject/viewid`,
      );
    });
  });
  describe('listViews()', () => {
    afterEach(() => {
      fetchMock.resetMocks();
    });

    it('should call the appropriate URL', async () => {
      fetchMock.mockResponse(JSON.stringify(mockEmptyList), { status: 200 });
      await listViews('myorg', 'myproject');
      expect(fetchMock.mock.calls[0][0]).toEqual(
        `${baseUrl}/views/myorg/myproject`,
      );
    });
    it('should return a Sparql and Elastic views', async () => {
      fetchMock.mockResponse(JSON.stringify(mockList), { status: 200 });
      const views = await listViews('myorg', 'myproject');
      expect(views.length).toEqual(2);
      expect(views).toMatchSnapshot();
    });
  });
  describe('getElasticSearchView()', () => {
    afterEach(() => {
      fetchMock.resetMocks();
    });

    it('should call the appropriate URL', async () => {
      fetchMock.mockResponse(JSON.stringify(mockElasticViewResponse), {
        status: 200,
      });
      await getElasticSearchView('myorg', 'myproject');
      expect(fetchMock.mock.calls[0][0]).toEqual(
        `${baseUrl}/views/myorg/myproject/nxv:defaultElasticSearchIndex`,
      );
    });
    it.skip('should throw', async () => {
      fetchMock.mockResponse(JSON.stringify(mockSparqlResponse), {
        status: 200,
      });
      expect(async () => {
        await getElasticSearchView('myorg', 'myproject');
      }).toThrow();
    });
  });
  describe('getSparqlView()', () => {
    afterEach(() => {
      fetchMock.resetMocks();
    });

    it('should call the appropriate URL', async () => {
      fetchMock.mockResponse(JSON.stringify(mockSparqlResponse), {
        status: 200,
      });
      await getSparqlView('myorg', 'myproject');
      expect(fetchMock.mock.calls[0][0]).toEqual(
        `${baseUrl}/views/myorg/myproject/nxv:defaultSparqlIndex`,
      );
    });
  });
});
