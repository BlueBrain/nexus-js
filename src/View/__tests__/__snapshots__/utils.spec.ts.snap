// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Views utils functions listViews() should return a Sparql and Elastic views 1`] = `
Array [
  ElasticSearchView {
    "deprecated": true,
    "id": "nxv:myview2",
    "orgLabel": "myorg",
    "projectLabel": "myproject",
    "queryURL": "/views/myorg/myproject/nxv%3Amyview2/_search",
    "rev": 4,
    "type": Array [
      "ElasticSearchView",
      "View",
      "Alpha",
    ],
    "uuid": "4f90ceff-45b7-442d-8536-d41705321d50",
  },
  SparqlView {
    "deprecated": true,
    "id": "nxv:myview",
    "orgLabel": "myorg",
    "projectLabel": "myproject",
    "queryURL": "/views/myorg/myproject/nxv%3Amyview/sparql",
    "rev": 4,
    "type": Array [
      "SparqlView",
      "View",
      "Alpha",
    ],
    "uuid": "4f90ceff-45b7-442d-8536-d41705321d50",
  },
]
`;
